### DISCUSSION TITLE
OCG response to other OCGs states

### DISCUSSION THREAD

**Discussion by claudinecombe** [pymupdf/PyMuPDF#2523] [Looking for help] [answered]
Hi 👋 

I am trying  to set some interactions among different OCGs, something like a "radiobutton" but with variation: check ocg1 and ocg2 when ocg3 is checked in the ui
Maybe the `add_layer() `function might help? Actually I have not figured out how to use it and what is the purpose.

Thank you for your help !

**Comment by claudinecombe** [comment]
I can add more concrete examples:

The first 4 checkbox will control the other ones
```
- [X] Custom display only
- [ ] Standard display only
- [ ] All
- [ ] Select
----
- [ ] Standard ocg1
- [ ] Standard ocg2
- [ ] Standard ocg3
- [X] Custom ocg4
```

```
- [ ] Custom display only
- [X] Standard display only
- [ ] All
- [ ] Select
----
- [X] Standard ocg1
- [X] Standard ocg2
- [X] Standard ocg3
- [ ] Custom ocg4
```

```
- [ ] Custom display only
- [ ] Standard display only
- [X] All
- [ ] Select
----
- [X] Standard ocg1
- [X] Standard ocg2
- [X] Standard ocg3
- [X] Custom ocg4
```

```
- [ ] Custom display only
- [ ] Standard display only
- [ ] All
- [X] Select # all checkboxes below are unchecked and the user may check the desired ones
----
- [ ] Standard ocg1
- [ ] Standard ocg2
- [ ] Standard ocg3
- [ ] Custom ocg4
```

**Reply by JorjMcKie** [reply]
I think you should use OCMD objects (Optional Content Membership Dictionary). These objects are logical expressions about the ON / OFF values of multiple OCGs.
The good thing is, that they allow practically arbitrary conditions via their `VE` ("visibility expression") parameter, see [here](https://pymupdf.readthedocs.io/en/latest/document.html#Document.set_ocmd) in the documentation. In the method description, there also is an example.
Other than that, an OCMD can be used like an OCG: it can be assigned to images, annotations, etc.
Compared to this, the radiobutton logic is a lot less flexible.

**Reply by claudinecombe** [reply]
Thank you Jorj for your reply !

I tried to use OCMDs and indeed it works very well, but they don't show up as checkboxes in the UI like OCGs, or maybe I missed something ?

**Reply by JorjMcKie** [reply]
Ah, I see your confusion now.
No, using the term "radiobutton group" in Optional Content context is something entirely different. Has nothing to do with radio buttons in Form Fields / Widgets.
The PDF specification only used the term to explain a roughly similar behavior: a radio button group is a list of OCG xrefs, of which no more than one OCG may be ON. Probably not the best idea.

**Reply by claudinecombe** [reply]
So there is no way to have such control with the checkboxes like in my examples above ? 
The OCMDs would work but it won't reflect in the checkboxes status

**Comment by JorjMcKie** [comment, marked as answer]
> So there is no way to have such control with the checkboxes like in my examples above ? The OCMDs would work but it won't reflect in the checkboxes status

This is correct - unfortunately. You can of course make your own checkboxes using `page.add_widget()` with appropriate logic in your script. You have to provide code that translates the checkbox widget value to the right `ve` syntax in `doc.set_ocmd(ve=[...])` ... and vice versa.

**Comment by claudinecombe** [comment]
Thank you Jorj, I was not aware of the `Widget` class, I will definitely take a look ! 👍

**Reply by JorjMcKie** [reply]
Ah, ok. This class allows you to add form fields. Among the several types are checkboxes. For how to add a checkbox, have a look at [this](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/fields/widgettest.py) script.
